README.md
# Healthchecks

Express middleware that runs health checks on your application.  Allows you to
manage a list of healthchecks as plain text files, keep them in source control
next to the application.  Provides a single endpoint you can access from a
mobile device, or from a monitoring service.


## What and Why?

A health check is a simple ping to a resource of your web application that
checks that your application is up and running, responding to network requests,
and behaving correctly.

It can be as simple as pinging the home page of a web site, checking that the
page includes the company's name in the title.

If you have a complex application, there are multiple things that can break
independently, and so you want a good health coverage by checking multiple
resources (see [What Should I Check?](#what-should-i-check)).

If your application has got one page that's accessing the database, and another
page that just storing requests in a queue, you want to check both[1].

Whether you're using a service like Pingdom or internal tool like Nagios, if
you store your checks there, funny thing is they never get updated when you roll
out new features.

You want checks to be part of the code base, in source control, right next to
the application code that's getting checked, where it can be versioned and code
reviewed.

And that's what this module does.  It lets you write your checks as a plain text
file that lives in the same repository as your application.

And it gives you a single endpoint that you can open in a browser, to see a list
of all passed or failed checks.  The same endpoint you can also use with a
monitoring service like Pingdom or Nagios.


## Checks File

The checks file lists one or more resources to check, and the expected content
of each resource (which may be empty).

For example:

```
# Check a web page
/                       My Amazing App

# Check stylesheets and scripts
/stylesheets/index.css  .body
/scripts/index.js       "use strict"

# Check the image exist
/images/logo.png
```

All URLs are relative to the server on which they are deployed, but must consist
of at least an absolute pathname.  You can include the hostname and protocol if
you need to, for example, if your tests are mounted at `http://example.com` and
you want to test `static.example.com` and test the HTTPS admin page:

```
# Check a different hostname than example.com
//static.example.com/logo.png

# Check with a different protocol and hostname
https://admin.example.com           Admin Dashboard
```

The expected content is matched literally against the body of the HTTP response.
Only 2xx responses are considered successful (however, redirects are followed).


## Usage

Install:

```bash
npm install --save healthchecks
```

Include the checks file with your web server, then configure the middleware to
read the checks file, for example:

```javascript
const healthchecks = require('healthchecks');

// This file contains all the checks
const CHECKS_FILE = __dirname + '/checks';

// Mount the middleware at /_healthchecks
server.use('/_healthchecks', healthchecks(CHECKS_FILE));
```

Now point your monitoring at `http://example.com/_healthchecks`.

You can also open this page with your browser to see a list of passing and
failed tests.

**Note** this endpoint is publicly accessible.  You can use another middleware
to add access control, or add this feature and send us a pull request.

You can initialize the middleware with the checks file name, or with an object
containing the following options:

* `filename`  -- The name of the checks file
* `onFailed`  -- Called with array of failed checks
* `timeout`   -- Timeout slow responses

You can specify the timeout in milliseconds or as a string, e.g. "3s" for 3
seconds.

Each failed check reported to `onFailed` is an object with the following
properties:

* `url`         -- The absolute URL
* `reason`      -- One of `'error'`, `'timeout'`, `'statusCode'` or `'body'`
* `error`       -- Connection or timeout error
* `timeout`     -- True if failed due to timeout
* `statusCode`  -- HTTP status code (if no error)
* `body`        -- Response body

For convenience, the value of the `reason` property is the name of one of the
other properties.  Also, when you call `toString()` you get a URL with the
reason, e.g. `"http://example.com => statusCode"`.

For example:

```javascript
const options = {
  filename:   CHECKS_FILE,
  timeout:    '5s',    // 5 seconds, can also pass duration in milliseconds
  onFailed:   function(checks) {
    checks.forEach(function(check) {
      log('The following check failed:', check.url, 'reason:', check.reason);
      // ... or ...
      log('The following check failed: %s', check);
    });
  }
};
server.use('/_healthchecks', healthchecks(options);
```


## What Should I Check?

> Anything that can go wrong will go wrong.
>
> -- Murphy's law

If two parts of your application can fail independently, you want to check both.

If you have static and dynamic page, you want to check both.

If different pages use different database servers, you want to check them all.

If some page uses a caching servers, you want to check that as well.

If another page uses a 3rd party API, also check that.

If your page is composed of multiple modules that can fail independently (say
shopping cart and product list), you want to check each module.

If something can fail that's not an HTML page, you want to check that as well.

Stylesheets? Check.  Client-side scripts?  Checks.  Images?  Checks.

If they are pre-processed, you want to check what was generated.

If it's served by a 3rd party CDN, don't skip this check.

If your application dynamically generates links (to pages, CSS, JS), check those
as well.

You can have too many checks, but most likely your problem is you don't have
enough!


## License

[MIT License](LICENSE) Copyright (c) 2014 Broadly Inc



CHANGELOG.md
## Version 1.7.3  2016-06-21

CHANGED code cleanups and linting


## Version 1.7.2  2015-12-15

ADDED checks for HTTPS URLs now set `X-Forwarded-Proto`


## Version 1.7.1  2015-12-09

FIXED erroneous package pushed to npm


## Version 1.7.0  2015-12-09

FIXED always run checks against `localAddress` and `localPort`

FIXED checking subdomains works again

CHANGED dropped short-lived support for SSL

CHANGED follow up to 10 redirects


## Version 1.6.1  2015-11-09

FIXED follow redirects when performing checks

ADDED skip SSL certificate validation via `strictSSL` option


## Version 1.5.0  2015-05-16

CHANGED show response time in high resolution (including nanoseconds)


## Version 1.4.4  2015-04-16

CHANGED show response time in seconds, 2 decimal places


## Version 1.4.3  2015-03-27

FIXED debug logging check requests


## Version 1.4.2  2015-03-27

FIXED connection.encrypted is now socket.encrypted (0.12/io.js)


## Version 1.4.1  2015-03-27

FIXED maintain the same connection protocol (in addition to IP/port)


## Version 1.4.0  2015-03-24

CHANGED run checks against localAddress/localPort


## Version 1.3.1  2015-03-23

ADDED send user-agent header, some servers require this


## Version 1.3.0  2015-03-23

CHANGED module can be used as Node request handler, does not need Express


## Version 1.2.5  2015-02-11

CHANGED better styling for elapsed time


## Version 1.2.4  2015-02-11

ADDED show request elapsed time


## Version 1.2.3  2015-01-19

CHANGED outcome.toString() no shows error message or status code


## Version 1.2.2  2015-01-12

FIXED verify result page shows consolidated results


## Version 1.2.1  2015-01-12

CHANGED consolidate multiple checks for the same URL

FIXED did not accept timeout option


## Version 1.2.0  2015-01-11

CHANGED options `onfailed` replaced with `onFailed` which accepts an array of
check results (not just URLs).

Each failed check reported to `onFailed` is an object with the following
properties:

`url`         -- The absolute URL
`reason`      -- One of 'error', 'timeout', 'statusCode' or 'body'
`error`       -- Connection or timeout error
`timeout`     -- True if failed due to timeout
`statusCode`  -- HTTP status code (if no error)
`body`        -- Response body


## Version 1.1.1  2014-12-08

CHANGED only show each URL once, and sort alphabetically


## Version 1.1.0  2014-12-03

ADDED onfailed option: called with list of failed checks

ADDED If the client sends an X-Request-Id header when making a healthcheck
request, that header is sent to all checked resources



test/user.js
const Browser = require('zombie');
const ms      = require('ms');
const server  = require('./server');


const checksURL = 'http://localhost:3000/_healthchecks';


describe('User', function() {

  const browser = Browser.create();

  before(function(done) {
    server.ready(done);
  });

  describe('runs checks, all healthy', function() {

    it('should see a list of passing tests', function(done) {
      browser.visit(checksURL, function() {
        browser.assert.text('h1', 'Passed');
        browser.assert.text('.passed li:nth-of-type(1)', 'http://localhost:3000/error');
        browser.assert.text('.passed li:nth-of-type(2)', 'http://localhost:3000/timeout');
        browser.assert.text('.passed li:nth-of-type(3)', 'http://localhost:3000/status');
        browser.assert.text('.passed li:nth-of-type(4)', 'http://localhost:3000/expected');
        done();
      });
    });
  });


  describe('runs checks, URL not accessible', function() {
    before(function() {
      server.locals.error = true;
    });

    it('should see a failed test', function(done) {
      browser.visit(checksURL, function() {
        browser.assert.text('h1', 'FailedPassed');
        browser.assert.elements('.failed li', 1);
        browser.assert.text('.failed li:nth-of-type(1)', 'http://localhost:3000/error');
        browser.assert.elements('.passed li', 3);
        done();
      });
    });

    after(function() {
      server.locals.error = false;
    });
  });


  describe('runs checks, response times out', function() {
    before(function() {
      server.locals.timeout = ms('3s');
    });

    it('should see a failed test', function(done) {
      this.timeout(ms('4s'));

      browser.visit(checksURL, function() {
        browser.assert.text('h1', 'FailedPassed');
        browser.assert.elements('.failed li', 1);
        browser.assert.text('.failed li:nth-of-type(1)', 'http://localhost:3000/timeout');
        browser.assert.elements('.passed li', 3);
        done();
      });
    });

    after(function() {
      server.locals.timeout = 0;
    });
  });


  describe('runs checks, response is 400', function() {
    before(function() {
      server.locals.status = 400;
    });

    it('should see a failed test', function(done) {
      browser.visit(checksURL, function() {
        browser.assert.text('h1', 'FailedPassed');
        browser.assert.elements('.failed li', 1);
        browser.assert.text('.failed li:nth-of-type(1)', 'http://localhost:3000/status');
        browser.assert.elements('.passed li', 3);
        done();
      });
    });

    after(function() {
      server.locals.status = 200;
    });
  });


  describe('runs checks, response missing expected content', function() {
    before(function() {
      server.locals.expected = 'Expected to see foo and also bar';
    });

    it('should see a failed test', function(done) {
      browser.visit(checksURL, function() {
        browser.assert.text('h1', 'FailedPassed');
        browser.assert.elements('.failed li', 1);
        browser.assert.text('.failed li:nth-of-type(1)', 'http://localhost:3000/expected');
        browser.assert.elements('.passed li', 3);
        done();
      });
    });

    after(function() {
      server.locals.expected = 'Expected to see foo and bar';
    });
  });


});



test/server.js
const express     = require('express');
const healthcheck = require('..');


const server = express();
module.exports = server;


// We have two test suites that want to run the same test server.
//
// Instead of calling listen() they call ready(), and get notified when the
// server is ready to receive requests.  Server only started once.
var listening = false;
server.ready = function(callback) {
  if (listening) {
    setImmediate(callback);
  } else {
    server.listen(3000, function() {
      listening = true;
      callback();
    });
  }
};


server.use('/_healthchecks', healthcheck({ filename: __dirname + '/checks' }));

// Test the response errors
server.locals.error = false;
server.get('/error', function(req, res) {
  if (server.locals.error)
    res.socket.destroy();
  else
    res.status(204).send('');
});

// Test the response timeout
server.locals.timeout = 0;
server.get('/timeout', function(req, res) {
  setTimeout(function() {
    res.send('');
  }, server.locals.timeout);
});

// Test the response has specific status code
server.locals.status = 200;
server.get('/status', function(req, res) {
  res.status(server.locals.status).send('');
});

// Test the response contains expected text
server.locals.expected = 'Expected to see foo and bar';
server.get('/expected', function(req, res) {
  res.send(server.locals.expected);
});


test/pingdom.js
const assert  = require('assert');
const ms      = require('ms');
const request = require('request');
const server  = require('./server');


const checksURL = 'http://localhost:3000/_healthchecks';


describe('Pingdom', function() {

  before(function(done) {
    server.ready(done);
  });


  describe('runs checks, all healthy', function() {

    it('should receive response with status 200', function(done) {
      request(checksURL, function(error, response) {
        assert.equal(response.statusCode, 200);
        done();
      });
    });
  });


  describe('runs checks, URL not accessible', function() {
    before(function() {
      server.locals.error = true;
    });

    it('should receive response with status 500', function(done) {
      request(checksURL, function(error, response) {
        assert.equal(response.statusCode, 500);
        done();
      });
    });

    after(function() {
      server.locals.error = false;
    });
  });


  describe('runs checks, response times out', function() {
    before(function() {
      server.locals.timeout = ms('3s');
    });

    it('should receive response with status 500', function(done) {
      this.timeout(ms('4s'));

      request(checksURL, function(error, response) {
        assert.equal(response.statusCode, 500);
        done();
      });
    });

    after(function() {
      server.locals.timeout = 0;
    });
  });


  describe('runs checks, response is 400', function() {
    before(function() {
      server.locals.status = 400;
    });

    it('should receive response with status 500', function(done) {
      request(checksURL, function(error, response) {
        assert.equal(response.statusCode, 500);
        done();
      });
    });

    after(function() {
      server.locals.status = 200;
    });
  });


  describe('runs checks, response missing expected content', function() {
    before(function() {
      server.locals.expected = 'Expected to see foo and also bar';
    });

    it('should receive response with status 500', function(done) {
      request(checksURL, function(error, response) {
        assert.equal(response.statusCode, 500);
        done();
      });
    });

    after(function() {
      server.locals.expected = 'Expected to see foo and bar';
    });
  });


});


lib/healthcheck.js
// Exports an Express middleware factory.
//
// Example:
//
//   const healthcheck = require('healthcheck');
//   const CHECKS_FILE = './checks';
//
//   server.use('/_healthchecks', healthcheck({ filename: CHECKS_FILE }));

const assert      = require('assert');
const Promise     = require('bluebird');
const debug       = require('debug')('healthcheck');
const File        = require('fs');
const Handlerbars = require('handlebars');
const ms          = require('ms');
const request     = require('request');
const URL         = require('url');


// Default timeout for checks; a slow server is a failed server.
const DEFAULT_TIMEOUT = ms('3s');


// Read the checks file and returns a check function (see checkFunction).
function readChecks(options) {
  const checks = File.readFileSync(options.filename, 'utf-8')
    .split(/[\n\r]+/)                 // Split into lines
    .map(function(line) {             // Ignore leading/trailing spaces
      return line.trim();
    })
    .filter(function(line) {          // Ignore empty lines
      return line.length;
    })
    .filter(function(line) {          // Ignore comments
      return line[0] !== '#';
    })
    .filter(function(line) {          // Ignore name = value pairs
      return !/^\w+=/.test(line);
    })
    .map(function(line) {             // Split line to URL + expected value
      const match = line.match(/^(\S+)\s*(.*)/);
      debug('Added check', match[1], match[2]);
      return {
        url:      match[1],
        expected: match[2]
      };
    })
    .map(function(check) {            // Valid URLs only
      // URLs may be relative to the server, so contain an absolute path
      const url = URL.parse(check.url);
      assert(url.pathname && url.pathname[0] === '/', 'Check URL must have absolute pathname');
      assert(!url.protocol || /^https?:$/.test(url.protocol), 'Check URL may only use HTTP/S protocol');
      return check;
    });

  // Returns a check function that will use these checks / settings
  const context = {
    checks:   checks,
    timeout:  options.timeout || DEFAULT_TIMEOUT
  };
  return checkFunction.bind(context);
}


// The check function will run all checks in parallel, and resolve to an object
// with the properties:
// passed  - A list of all check URLs that passed
// failed  - A list of all check URLs that failed
function checkFunction(baseURL) {
  const checks  = this.checks;
  const timeout = this.timeout;

  // Checks have relative URLs, resolve them to absolute URLs
  // We use this array twice, when mapping to promises that check the URL
  // (allChecks), and when reducing check results to aggregate (passedAndFailed)
  const urls    = checks.map(function(check) {
    return URL.resolve(baseURL, check.url);
  });

  // Each check resolves into a boolean: success or failure
  const allChecks = checks.map(function(check, index) {
    return new Promise(function(resolve) {
      const url = urls[index];
      request({ uri: url, timeout: timeout }, function(error, response, body) {
        if (error) {
          debug('Server responded with error ' + error.message, url);
          resolve(false);
        } else if (response.statusCode < 200 || response.statusCode >= 400) {
          debug('Server responded with status code ' + response.statusCode, url);
          resolve(false);
        } else if (body.indexOf(check.expected) < 0) {
          debug('Server response did not contain expected text', url);
          resolve(false);
        } else
          resolve(true);
      });
    });
  });

  // Run all checks in parallel
  const allResults = Promise.all(allChecks);
 
  // Reduce into an object with the passed and failed lists of URLs
  const passedAndFailed = allResults.then(function(results) {
    // Partition the results into an array of passed and failed URLs
    const initial = {
      passed: [],
      failed: []
    };
    return results.reduce(function(totals, passed, index) {
      const url = urls[index];
      if (passed)
        totals.passed.push(url);
      else
        totals.failed.push(url);
      return totals;
    }, initial);
  });

  // Returns the promise
  return passedAndFailed;
}


// Call this function to configure and return the middleware.
module.exports = function healthcheck(options) {
  assert(options, 'Missing options');
  assert(options.filename, 'Missing checks filename');

  // Read all checks form the file and returns a checking function
  const runChecks = readChecks(options);

  // Load Handlebars template for rendering results
  const template  = File.readFileSync(__dirname + '/results.hbs', 'utf-8');
  const render    = Handlerbars.compile(template);

  // Return the Express middleware
  return function(request, response) {
    debug('Health checks: running');

    // The base URL where this middleware is mounted on.  We do this
    // dynamically, so sending healthchecks request to HTTP vs HTTPS, or
    // different domain, may result in running different checks.
    const mounted = URL.format({
      protocol: request.protocol,
      host:     request.headers.host,
      port:     request.port
    });

    // Run all checks
    runChecks(mounted)
      .then(function(results) {
        debug('Health checks: ' + results.passed.length + ' passed and ' + results.failed.length + ' failed');

        // Respond with 200 only if all checks passed
        // Respond with 500 if any check fail, or there are no checks
        const statusCode  = (results.passed.length === 0 || results.failed.length) ? 500 : 200;
        // Render template
        const html        = render(results);
        response.status(statusCode).send(html);
      });
  };

};


lib/healthchecks.js
'use strict';
// Exports an Express middleware factory.
//
// Example:
//
//   const healthchecks = require('healthchecks');
//   const CHECKS_FILE = './checks';
//
//   server.use('/_healthchecks', healthchecks(CHECKS_FILE));
//
// If you want to change the check timeout, use named argument:
//
//   const options = {
//     filename: CHECKS_FILE,
//     timeout:  '5s'     // 5 seconds, can also pass duration in milliseconds
//   };
//   server.use('/_healthchecks', healthchecks(options);


const assert     = require('assert');
const debug      = require('debug')('healthchecks');
const File       = require('fs');
const Handlebars = require('handlebars');
const hrTime     = require('pretty-hrtime');
const HTTP       = require('http');
const ms         = require('ms');
const Path       = require('path');
const URL        = require('url');


// Default timeout for checks; a slow server is a failed server.
const DEFAULT_TIMEOUT = '3s';

// HTTP status codes which we follow as redirects.
const REDIRECT_STATUSES = [301, 302, 303, 307];

// The maximum amount of redirects we will follow.
const MAX_REDIRECT_COUNT = 10;


// Represents a check outcome with the properties url, reason, etc.
//
// url        - Checked URL
// elapsed    - In milliseconds
// error      - Protocol error
// statusCode - HTTP status code
// body       - Response body
// expected   - Expected response body (array)
function Outcome(args) {
  this.url      = args.url;
  this.elapsed  = hrTime(args.elapsed);
  if (args.error && args.error.code === 'ETIMEDOUT') {
    this.reason   = 'timeout';
    this.timeout  = true;
  } else if (args.error) {
    this.reason   = 'error';
    this.error    = args.error;
  } else {

    this.statusCode = args.statusCode;
    this.body       = args.body;
    if (this.statusCode < 200 || this.statusCode >= 400)
      this.reason   = 'statusCode';
    else {

      const expected    = args.expected;
      const allMatching = expected.every(text => ~args.body.indexOf(text));
      if (!allMatching)
        this.reason = 'body';

    }
  }
  this.log();
}


// Log outcome, visible when DEBUG=healthchecks
Outcome.prototype.log = function() {
  switch (this.reason) {
    case 'error': {
      debug('%s: Server responded with error %s', this.url, this.error);
      break;
    }
    case 'timeout': {
      debug('%s: Server response timeout', this.url);
      break;
    }
    case 'statusCode': {
      break;
    }
    case 'body': {
      debug('%s: Server response did not contain expected text', this.url);
      break;
    }
    // no default
  }
};


// Easy way to log check outcome, shows the URL and the reason check failed
Outcome.prototype.toString = function() {
  switch (this.reason) {
    case 'error': {
      return `${this.url} => ${this.error.message}`;
    }
    case 'statusCode': {
      return `${this.url} => ${this.statusCode}`;
    }
    case undefined: {
      debug('%s: Server responded with status code %s', this.url, this.statusCode);
      return this.url;
    }
    default: {
      return `${this.url} => ${this.reason}`;
    }
  }
};


// Our Handlerbars instance.
const handlebars = Handlebars.create();


// Checks the status of the given relative URL.
// This URL, and all possible redirects, are resolved using
// the resolver function.
function runCheck(args) {
  const url       = args.url;
  const headers   = args.headers;
  const resolver  = args.resolver;
  const timeout   = args.timeout;
  const redirects = args.redirects || 0;

  const loopbackURL = resolver(url);
  const request     = {
    host:     loopbackURL.host,
    port:     loopbackURL.port,
    path:     loopbackURL.path,
    headers:  Object.assign({}, headers, {
      'Host':              loopbackURL.hostname,
      'X-Forwarded-Proto': loopbackURL.protocol.replace(/:$/, '')
    })
  };

  return get(request, timeout)
    .then(function(response) {
      const isRedirect = ~REDIRECT_STATUSES.indexOf(response.statusCode);
      if (isRedirect) {

        const redirectLoop = redirects >= MAX_REDIRECT_COUNT;
        if (redirectLoop)
          throw new Error('too many redirects');
        else {
          const redirectURL = response.headers.location;
          const sameDomain  = withinSameDomain(resolver(redirectURL), loopbackURL);
          if (sameDomain)
            return runCheck({ url: redirectURL, headers, resolver, timeout, redirects: redirects + 1 });
          else
            return response;
        }

      } else
        return response;
    });
}


// Makes an promisified HTTP GET request.
//
// On success, promise is resolved with { statusCode, headers, body }.
// On error or timeout, promise is rejected with error.
//
function get(request, timeout) {
  return new Promise(function(resolve, reject) {
    HTTP.get(request)
      .on('error', reject)
      .on('response', function(response) {
        const buffers = [];
        response.on('data', function(buffer) {
          buffers.push(buffer);
        });
        response.on('end', function() {
          const body = Buffer.concat(buffers).toString();
          resolve({
            statusCode: response.statusCode,
            headers:    response.headers,
            body:       body
          });
        });
      });

    setTimeout(function() {
      const error = new Error('ETIMEDOUT');
      error.code  = 'ETIMEDOUT';
      reject(error);
    }, timeout);
  });
}


function withinSameDomain(from, to) {
  const fromHost = from.hostname;
  const toHost   = to.hostname;
  return fromHost === toHost ||
         fromHost.endsWith(`.${toHost}`) ||
         toHost.endsWith(`.${fromHost}`);
}


// The check function will run all checks in parallel, and resolve to an object
// with the properties:
// passed  - A list of all check URLs that passed
// failed  - A list of all check URLs that failed
function checkFunction(url, requestID) {
  const protocol = URL.parse(url).protocol;
  const hostname = URL.parse(url).hostname;
  const port     = URL.parse(url).port;

  const checks  = this.checks;
  const timeout = this.timeout;

  // Given a relative URL in string form, returns a parsed URL
  // which points to the local server's IP address and port
  // and has the right hostname property to use in the Host header.
  function loopbackResolve(relativeURL) {
    const absoluteURL = URL.parse(URL.resolve(`${protocol}//localhost/`, relativeURL));
    const loopbackURL = Object.assign({}, absoluteURL, {
      hostname: absoluteURL.hostname,
      host:     hostname,
      port:     port
    });
    return loopbackURL;
  }

  // Each check resolves into an outcome object
  const allChecks = Object.keys(checks).map(function(checkURL) {
    const expected = checks[checkURL];

    // We need to make an HTTP/S request to the current server,
    // based on the hostname/port passed to us,
    // so the HTTP check would go to http://localhost:80/ or some such URL.
    const headers = {
      'User-Agent':   'Mozilla/5.0 (compatible) Healthchecks http://broadly.com',
      'X-Request-Id': requestID || ''
    };

    const start = process.hrtime();
    return runCheck({ url: checkURL, headers, resolver: loopbackResolve, timeout })
      .then(function(response) {
        const elapsed = process.hrtime(start);
        const outcome = new Outcome({ url: checkURL, expected, statusCode: response.statusCode, body: response.body, elapsed });
        return outcome;
      })
      .catch(function(error) {
        const elapsed = process.hrtime(start);
        const outcome = new Outcome({ url: checkURL, expected, error, elapsed });
        return outcome;
      });
  });


  // Run all checks in parallel
  const allOutcomes = Promise.all(allChecks);

  // Reduce into an object with the passed and failed lists of URLs
  const passedAndFailed = allOutcomes
    .then(function(outcomes) {
      return {
        passed: outcomes.filter(outcome => !outcome.reason ),
        failed: outcomes.filter(outcome => outcome.reason )
      };
    });

  // Returns the promise
  return passedAndFailed;
}


// Read the checks file and returns a check function (see checkFunction).
function readChecks(filename, timeout) {
  const checks = File.readFileSync(filename, 'utf-8')
    .split(/[\n\r]+/)                     // Split into lines
    .map(line => line.trim())             // Ignore leading/trailing spaces
    .filter(line => line.length)          // Ignore empty lines
    .filter(line => line[0] !== '#')      // Ignore comments
    .filter(line => !/^\w+=/.test(line))  // Ignore name = value pairs
    .map(function(line) {             // Split line to URL + expected value
      const match = line.match(/^(\S+)\s*(.*)/);
      return {
        url:      match[1],
        expected: match[2]
      };
    })
    .map(function(check) {            // Valid URLs only
      // URLs may be relative to the server, so contain an absolute path
      const url = URL.parse(check.url);
      assert(url.pathname && url.pathname[0] === '/', 'Check URL must have absolute pathname');
      assert(!url.protocol || /^https?:$/.test(url.protocol), 'Check URL may only use HTTP/S protocol');
      return check;
    })
    .reduce(function(memo, check) {
      const url = check.url;
      memo[url] = memo[url] || []; // eslint-disable-line
      if (check.expected)
        memo[url].push(check.expected);
      return memo;
    }, {});

  // Returns a check function that will use these checks / settings
  const context = {
    checks:   checks,
    timeout:  timeout
  };
  debug('Added %d checks', checks.length);

  return checkFunction.bind(context);
}


// Returns a comparer function suitable for Array.sort().
function compareProperty(propName) {
  return function(a, b) {
    if (a[propName] < b[propName])
      return -1;
    if (a[propName] > b[propName])
      return 1;
    return 0;
  };
}


// Respond with 200 only if all checks passed
// Respond with 500 if any check fail
// Respond with 404 if there are no checks to run
function statusCodeFromOutcomes(passed, failed) {
  const anyFailed = failed.length > 0;
  const anyPassed = passed.length > 0;
  if (anyFailed)
    return 500;
  else if (anyPassed)
    return 200;
  else
    return 404;
}


// Call this function to configure and return the middleware.
module.exports = function healthchecks(options) {
  assert(options, 'Missing options');

  // Pass filename as first argument or named option
  const filename    = typeof (options) === 'string' ? options : options.filename;
  assert(filename, 'Missing checks filename');

  // Pass timeout as named option, or use default
  const timeoutArg  = (typeof (options) === 'object' && options.timeout) || DEFAULT_TIMEOUT;
  // If timeout argument is a string (e.g. "3d"), convert to milliseconds
  const timeout     = (typeof (timeoutArg) === 'string') ? ms(timeoutArg) : parseInt(timeoutArg, 10);

  const onFailed    = options.onFailed || function() {};

  // Read all checks form the file and returns a checking function
  const runChecks   = readChecks(filename, timeout);

  // Load Handlebars template for rendering results
  const template    = File.readFileSync(Path.join(__dirname, '/index.hbs'), 'utf-8');
  const render      = handlebars.compile(template);


  // Return the Express middleware
  return function(req, res) {

    const requestID = req.headers['x-request-id'];

    // We use local address/port to health check this server, e.g. the checks
    // may say //www.example.com/ but in development we connect to
    // 127.0.0.1:5000
    const protocol  = req.socket.encrypted ? 'https:' : 'http:';
    const hostname  = req.socket.localAddress;
    const port      = req.socket.localPort;
    const url       = URL.format({ protocol, hostname, port });

    // Run all checks
    debug('Running against %s://%s:%d with request-ID %s', protocol, hostname, port, requestID);
    runChecks(url, requestID)
      .then(function(outcomes) {
        debug('%d passed and %d failed', outcomes.passed.length, outcomes.failed.length);

        const passed      = outcomes.passed.sort(compareProperty('url'));
        const failed      = outcomes.failed.sort(compareProperty('url'));
        const statusCode  = statusCodeFromOutcomes(passed, failed);

        const html        = render({ passed, failed });
        res.writeHeader(statusCode);
        res.write(html);
        res.end();

        if (failed.length > 0)
          onFailed(failed);
      });
  };

};


